File: GitToolBox/src/main/java/zielu/gittoolbox/completion/CompletionCheckinHandler.java
package zielu.gittoolbox.completion;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.vcs.CheckinProjectPanel;
import com.intellij.openapi.vcs.VcsException;
import com.intellij.openapi.vcs.checkin.CheckinHandler;
import com.intellij.ui.EditorTextComponent;
import java.io.File;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import javax.swing.JComponent;
import org.jetbrains.annotations.NotNull;
import zielu.gittoolbox.metrics.ProjectMetrics;

class CompletionCheckinHandler extends CheckinHandler implements CompletionScopeProvider {
  private final CheckinProjectPanel panel;

  CompletionCheckinHandler(CheckinProjectPanel panel) {
    this.panel = panel;
  }

  @Override
  public void includedChangesChanged() {
    //do nothing
  }

  @Override
  public void checkinSuccessful() {
    //do nothing
  }

  @Override
  public void checkinFailed(List<VcsException> exception) {
    //do nothing
  }

  @NotNull
  @Override
  public Collection<File> getAffectedFiles() {
    if (isCommitUiActive()) {
      Project project = panel.getProject();
      return ProjectMetrics.getInstance(project).timer("completion-get-affected")
          .timeSupplier(panel::getFiles);
    }
    return Collections.emptyList();
  }

  private boolean isCommitUiActive() {
    JComponent component = panel.getPreferredFocusedComponent();
    if (component instanceof EditorTextComponent) {
      return ((EditorTextComponent) component).getComponent().isFocusOwner();
    }
    return false;
  }
}

File: src/main/kotlin/zielu/gittoolbox/completion/CompletionCheckinHandler.kt
package zielu.gittoolbox.completion

import com.intellij.openapi.vcs.CheckinProjectPanel
import com.intellij.openapi.vcs.checkin.CheckinHandler
import com.intellij.ui.EditorTextComponent
import zielu.gittoolbox.metrics.ProjectMetrics
import java.io.File

internal class CompletionCheckinHandler(
  private val panel: CheckinProjectPanel
) : CheckinHandler(), CompletionScopeProvider {
  override fun getAffectedFiles(): Collection<File> {
    if (isCommitUiActive()) {
      val project = panel.project
      return ProjectMetrics.getInstance(project).timer("completion-get-affected")
        .timeSupplierKt { panel.files }
    }
    return emptyList()
  }

  private fun isCommitUiActive(): Boolean {
    val component = panel.preferredFocusedComponent
    return if (component is EditorTextComponent) {
      (component as EditorTextComponent).component.isFocusOwner
    } else false
  }
}
